# Introduction

Have you ever thought about building your own iOS or Android app? If so, then you probably have wondered about how things work in the app stores. Today we'll replicate some of the app store analytics provided by companies like App Annie or Sensor Tower that helps inform development and app marketing strategies for many companies. This stuff is BIG business!

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/datasets/lava18/google-play-store-apps).

# Objectives

In this module, we will compare thousands of apps in the Google Play Store so that we can gain insight into:

- How competitive different app categories (e.g., Games, Lifestyle, Weather) are
- Which app category offers compelling opportunities based on its popularity
- How many downloads you would give up by making your app paid vs. free
- How much you can reasonably charge for a paid app
- Which paid apps have had the highest revenue
- How many paid apps will recoup their development costs based on their sales revenue

# Import Statements

In [198]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import seaborn as sns
import matplotlib.pyplot as plt


# Notebook Presentation

In [134]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [135]:
df_apps = pd.read_csv('apps.csv')

# Data Cleaning

In [136]:
# Display the number of rows and columns
print(f"Data Shape: {df_apps.shape}\n\n")
print("Data Types", df_apps.dtypes)




Data Shape: (10841, 12)


Data Types App                object
Category           object
Rating            float64
Reviews             int64
Size_MBs          float64
Installs           object
Type               object
Price              object
Content_Rating     object
Genres             object
Last_Updated       object
Android_Ver        object
dtype: object


In [137]:
# Convert the Last Updated to Datetime Format
df_apps['Last_Updated'] = pd.to_datetime(df_apps['Last_Updated'])


In [138]:
# Display sample rows
print(f"Column Names: {df_apps.columns.tolist()}")

Column Names: ['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type', 'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Android_Ver']


In [139]:
# Display first few rows
print("Data Sample:\n")
df_apps.head(5)

Data Sample:



App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  \
0        0      8.70        0  Paid  $13.99           Teen           Social   
1        0     33.00        0  Paid   $2.99       Everyone        Education   
2        0      5.50        0  Paid   $1.49       Everyone  Personalization   
3        0     19.00        0   NaN       0   Everyone 10+         Strategy   
4        0     10.00        0  Free       0       Everyone         Business   

  Last_Updated         Android_Ver  
0   2017-07-28          4.1 and up  
1   2016-04-15          3.0 and up  
2   2018-07-11          4.2 and up  
3   2018-06-28  Varies with device  
4   2018-08-06          4.1 and up

In [140]:
# # Drop unwanted columns from the dataset
# df_apps_cleaned = df_apps.drop(columns=['Last_Updated', 'Android_Ver'])
# df_apps_cleaned.head(5)


## FInd and Remove Rows with null values

In [141]:
# Find number of rows with missing values
df_apps_cleaned.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size_MBs          0
Installs          0
Type              0
Price             0
Content_Rating    0
Genres            0
Last_Updated      0
Android_Ver       0
dtype: int64

In [142]:
# Drop rows with missing values and verify the result
df_apps_cleaned = df_apps.dropna(axis=0)
df_apps_cleaned.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size_MBs          0
Installs          0
Type              0
Price             0
Content_Rating    0
Genres            0
Last_Updated      0
Android_Ver       0
dtype: int64

### Find and Remove Duplicates




In [143]:
# Find columns with duplicated values
duplicated_cols = [col for col in df_apps_cleaned.columns if df_apps_cleaned[col].duplicated().any()]
print(f"Columns with Duplicated Columns: {duplicated_cols}\n")

# Find rows with duplicated rows
print("Rows with Duplicated Rows:", int(df_apps_cleaned.duplicated().sum()))


Columns with Duplicated Columns: ['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type', 'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Android_Ver']

Rows with Duplicated Rows: 474


In [144]:
df_apps_cleaned[df_apps_cleaned.duplicated()]

App            Category  Rating  \
946                        420 BZ Budeze Delivery             MEDICAL    5.00   
1133                                  MouseMingle              DATING    2.70   
1196   Cardiac diagnosis (heart rate, arrhythmia)             MEDICAL    4.40   
1231                                 Sway Medical             MEDICAL    5.00   
1247               Chat Kids - Chat Room For Kids              DATING    4.70   
...                                           ...                 ...     ...   
10802               Skype - free IM & video calls       COMMUNICATION    4.10   
10809                                   Instagram              SOCIAL    4.50   
10826                                Google Drive        PRODUCTIVITY    4.40   
10832                                 Google News  NEWS_AND_MAGAZINES    3.90   
10839                              Subway Surfers                GAME    4.50   

        Reviews  Size_MBs       Installs  Type   Price Content_Rating  \
946           2     11.00            100  Free       0     Mature 17+   
1133          3      3.90            100  Free       0     Mature 17+   
1196          8      6.50            100  Paid  $12.99       Everyone   
1231          3     22.00            100  Free       0       Everyone   
1247          6      4.90            100  Free       0     Mature 17+   
...         ...       ...            ...   ...     ...            ...   
10802  10484169      3.50  1,000,000,000  Free       0       Everyone   
10809  66577313      5.30  1,000,000,000  Free       0           Teen   
10826   2731211      4.00  1,000,000,000  Free       0       Everyone   
10832    877635     13.00  1,000,000,000  Free       0           Teen   
10839  27725352     76.00  1,000,000,000  Free       0   Everyone 10+   

                 Genres Last_Updated         Android_Ver  
946             Medical   2018-06-06          4.1 and up  
1133             Dating   2018-07-17          4.4 and up  
1196            Medical   2018-07-25          3.0 and up  
1231            Medical   2018-07-25          5.0 and up  
1247             Dating   2018-07-24        4.0.3 and up  
...                 ...          ...                 ...  
10802     Communication   2018-08-03  Varies with device  
10809            Social   2018-07-31  Varies with device  
10826      Productivity   2018-08-06  Varies with device  
10832  News & Magazines   2018-08-01          4.4 and up  
10839            Arcade   2018-07-12          4.1 and up  

[474 rows x 12 columns]

In [146]:
# Remove duplicated rows and verify
df_apps_cleaned = df_apps_cleaned.drop_duplicates(keep='first')
print("Number of Duplicated Rows:", int(df_apps_cleaned.duplicated().sum()))

# Check whether the same app appears multiple times
df_apps_cleaned[df_apps_cleaned['App'].duplicated()]

# Keep the latest app entry
# Remove duplicates by keeping the latest date per App
df_apps_cleaned = df_apps_cleaned.loc[df_apps_cleaned.groupby('App')['Last_Updated'].idxmax()].reset_index(drop=True)





Number of Duplicated Rows: 0


# Find Highest Rated Apps



In [147]:
# Find Top Rated Apps
df_apps_cleaned.sort_values(by="Rating", ascending=False).head(5)


App            Category  Rating  Reviews  \
1717               CP Installer App            BUSINESS    5.00        4   
1704  COMSATS BOOK STORE FOR BS(CS)              FAMILY    5.00       15   
7053                  The CJ Rubric  HEALTH_AND_FITNESS    5.00        5   
5412          NOMISMA.com.cy by FMW  NEWS_AND_MAGAZINES    5.00        3   
1697                    CN Resident              FAMILY    5.00        1   

      Size_MBs Installs  Type Price Content_Rating            Genres  \
1717     24.00      100  Free     0       Everyone          Business   
1704     94.00       50  Free     0       Everyone         Education   
7053     43.00      100  Free     0       Everyone  Health & Fitness   
5412      6.30      100  Free     0       Everyone  News & Magazines   
1697     64.00      100  Free     0       Everyone     Entertainment   

     Last_Updated   Android_Ver  
1717   2018-07-24    4.1 and up  
1704   2018-07-11  4.0.3 and up  
7053   2017-12-05    4.1 and up  
5412   2016-05-29    4.0 and up  
1697   2017-05-22    4.1 and up

In [148]:
# Find all the apps with a Rating of '5'
highest_rated_apps = df_apps_cleaned[df_apps_cleaned['Rating'] == 5]
print("Number of apps with the highest rating:", highest_rated_apps.shape[0])

Number of apps with the highest rating: 271


In [149]:
# Find the category of apps with the highest rating
highest_rated_apps['Category'].value_counts(normalize=True) * 100

Category
FAMILY                24.72
LIFESTYLE             10.70
MEDICAL                9.23
BUSINESS               6.64
TOOLS                  6.27
HEALTH_AND_FITNESS     4.43
GAME                   4.43
PERSONALIZATION        3.69
FINANCE                2.95
SOCIAL                 2.95
PRODUCTIVITY           2.95
NEWS_AND_MAGAZINES     2.58
SHOPPING               2.21
PHOTOGRAPHY            2.21
DATING                 2.21
EVENTS                 2.21
BOOKS_AND_REFERENCE    2.21
COMMUNICATION          1.85
SPORTS                 1.48
TRAVEL_AND_LOCAL       1.11
FOOD_AND_DRINK         0.74
COMICS                 0.74
LIBRARIES_AND_DEMO     0.74
PARENTING              0.37
ART_AND_DESIGN         0.37
Name: proportion, dtype: float64

# Find 5 Largest Apps in terms of Size (MBs)



In [150]:
#Filter the largest apps by size

largest_apps_by_size = df_apps_cleaned.sort_values(by='Size_MBs', ascending=False)
print("Top 5 largest apps by size:")
largest_apps_by_size[['App', 'Size_MBs']].head(5).reset_index(drop=True)

Top 5 largest apps by size:


App  Size_MBs
0   Stickman Legends: Shadow Wars    100.00
1                 SimCity BuildIt    100.00
2  Talking Babsy Baby: Baby Games    100.00
3      Draft Simulator for FUT 18    100.00
4                       Post Bank    100.00

# Find the 5 App with Most Reviews



In [152]:
apps_most_reviews = df_apps_cleaned.sort_values(by='Reviews', ascending=False)
apps_most_reviews[['App', 'Reviews']].head(5).reset_index(drop=True)



App   Reviews
0                                  Facebook  78158306
1                        WhatsApp Messenger  69119316
2                                 Instagram  66577313
3  Messenger – Text and Video Chat for Free  56642847
4                            Clash of Clans  44891723

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [155]:
# Count of apps per rating
fig = px.pie(df_apps_cleaned,
             names='Content_Rating',
             title='Distribution of Content Ratings',
             hole=0.4)  # hole=0.4 makes it a Donut chart

fig.show()

# Numeric Type Conversion: Examine the Number of Installs


In [157]:
# Verify the data type of the 'Installs' column
df_apps_cleaned['Installs'].dtype



dtype('O')

In [163]:
# Remove commas and plus signs
df_apps_cleaned['Installs_numeric'] = df_apps_cleaned['Installs'].str.replace(',', '').str.replace('+', '', regex=False)

# Convert to integer
df_apps_cleaned['Installs_numeric'] = pd.to_numeric(df_apps_cleaned['Installs_numeric'])

# Verify the change
df_apps_cleaned['Installs_numeric'].dtype


dtype('int64')

In [164]:
# Number of installs
df_apps_cleaned['Installs_numeric'].value_counts().sort_index()


Installs_numeric
1                3
5                9
10              69
50              56
100            302
500            199
1000           696
5000           425
10000          986
50000          457
100000        1094
500000         503
1000000       1416
5000000        607
10000000       936
50000000       204
100000000      189
500000000       24
1000000000      20
Name: count, dtype: int64

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate





In [165]:
# Clean the Price Column and convert to numeric
df_apps_cleaned['Price'] = df_apps_cleaned['Price'].str.replace('$', '', regex=False)
df_apps_cleaned['Price'] = pd.to_numeric(df_apps_cleaned['Price'], errors='coerce')


### The most expensive apps

In [172]:
# Display the top 20 highest paid apps



# Remove the junk apps above $250
top_20_expensive = top_20_expensive = df_apps_cleaned[df_apps_cleaned['Price'] <= 250]
top_20_expensive = top_20_expensive.sort_values('Price', ascending=False)[['App', 'Category', 'Price']].head(20).reset_index(drop=True)
top_20_expensive


App     Category  Price
0                           Vargo Anesthesia Mega App      MEDICAL  79.99
1                                        LTC AS Legal      MEDICAL  39.99
2                                    I am Rich Person    LIFESTYLE  37.99
3                             A Manual of Acupuncture      MEDICAL  33.99
4                                  PTA Content Master      MEDICAL  29.99
5                                            EMT PASS      MEDICAL  29.99
6                           AP Art History Flashcards       FAMILY  29.99
7                             Golfshot Plus: Golf GPS       SPORTS  29.99
8    Human Anatomy Atlas 2018: Complete 3D Human Body      MEDICAL  24.99
9   Muscle Premium - Human Anatomy, Kinesiology, B...      MEDICAL  24.99
10                         USMLE Step 2 CK Flashcards       FAMILY  19.99
11                               Hospitalist Handbook      MEDICAL  19.99
12                                         NewTek NDI  PHOTOGRAPHY  19.99
13                                  DRAGON QUEST VIII       FAMILY  19.99
14                            boattheory.ch Full 2018       FAMILY  19.40
15                             I am Rich Premium Plus      FINANCE  18.99
16                            The World Ends With You         GAME  17.99
17                                 SkyTest BU/GU Lite     BUSINESS  17.99
18                            Trine 2: Complete Story         GAME  16.99
19                         2017 EMRA Antibiotic Guide      MEDICAL  16.99

### Highest Grossing Paid Apps (ballpark estimate)

In [175]:
# Add a 'Revenue_Estimate' to calculate the approximate revenue earned
df_apps_cleaned['Revenue_Estimate'] = df_apps_cleaned['Price'] * df_apps_cleaned['Installs_numeric']

# Filter the top 10 apps with the highest revenue
top_10_revenue = df_apps_cleaned.sort_values(by='Revenue_Estimate', ascending=False).head(10)
top_10_revenue[['App', 'Price', 'Installs_numeric', 'Revenue_Estimate', 'Category']].reset_index(drop=True)

App  Price  Installs_numeric  Revenue_Estimate  \
0                      Minecraft   6.99          10000000     69,900,000.00   
1                      I am rich 399.99            100000     39,999,000.00   
2              I Am Rich Premium 399.99             50000     19,999,500.00   
3                  Hitman Sniper   0.99          10000000      9,900,000.00   
4  Grand Theft Auto: San Andreas   6.99           1000000      6,990,000.00   
5            Facetune - For Free   5.99           1000000      5,990,000.00   
6        Sleep as Android Unlock   5.99           1000000      5,990,000.00   
7            DraStic DS Emulator   4.99           1000000      4,990,000.00   
8       I'm Rich - Trump Edition 400.00             10000      4,000,000.00   
9                 I am Rich Plus 399.99             10000      3,999,900.00   

      Category  
0       FAMILY  
1    LIFESTYLE  
2      FINANCE  
3         GAME  
4         GAME  
5  PHOTOGRAPHY  
6    LIFESTYLE  
7         GAME  
8    LIFESTYLE  
9       FAMILY

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [179]:
# Count number of apps in each category
category_counts = df_apps_cleaned['Category'].value_counts().reset_index()
category_counts.columns = ['Category', 'Count']

# Plot bar chart
fig = px.bar(category_counts, x='Category', y='Count',
             title='Number of Apps per Category',
             text='Count')  # show values on bars
fig.update_layout(xaxis_title='Category', yaxis_title='Number of Apps')
fig.show()

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [180]:
downloads_by_category = df_apps_cleaned.groupby('Category')['Installs_numeric'].sum().sort_values(ascending=True).reset_index()


In [182]:
fig = px.bar(downloads_by_category[:10],
             x='Installs_numeric',
             y='Category',
             orientation='h',  # horizontal bars
             title='Most Popular Categories by Total Downloads',
             text='Installs_numeric')

fig.update_layout(xaxis_title='Total Downloads',
                  yaxis_title='Category')
fig.show()

### Category Concentration - Downloads vs. Competition


In [183]:
category_summary = df_apps_cleaned.groupby('Category').agg(
    Number_of_Apps=('App', 'count'),
    Total_Installs=('Installs_numeric', 'sum')
).reset_index()

In [184]:

fig = px.scatter(category_summary,
                 x='Number_of_Apps',
                 y='Total_Installs',
                 size='Total_Installs',      # optional: size by downloads
                 color='Category',           # optional: color by category
                 hover_name='Category',
                 title='Category Concentration: Downloads vs Competition',
                 labels={'Number_of_Apps': 'Number of Apps',
                         'Total_Installs': 'Total Downloads'}
                )

fig.update_layout(xaxis_title='Number of Apps (Competition)',
                  yaxis_title='Total Downloads (Popularity)')

fig.show()

# Extracting Nested Data from a Column




In [186]:
# Split genres by ';'
genres_split = df_apps_cleaned['Genres'].str.split(';')

# Convert the list of genres into a Series where each genre gets its own row
all_genres = genres_split.explode()

# Count the occurrences of each genre
genre_counts = all_genres.value_counts()
num_unique_genres = all_genres.nunique()
print("Number of unique genres:", num_unique_genres)


Number of unique genres: 53


# Grouped Bar Charts: Free vs. Paid Apps per Category

In [192]:
# Count number of free vs paid apps per category
category_type_counts = df_apps_cleaned.groupby(['Category', 'Type']).size().reset_index(name='Count')

fig = px.bar(category_type_counts,
             x='Category',
             y='Count',
             color='Type',          # Free vs Paid
             barmode='group',       # grouped bars
             title='Number of Free vs Paid Apps per Category')

fig.update_layout(xaxis_title='Category',
                  yaxis_title='Number of Apps')

fig.show()


# Plotly Box Plots: Lost Downloads for Paid Apps



In [193]:

fig = px.box(df_apps_cleaned,
             x='Type',                 # Free vs Paid
             y='Installs_numeric',     # Number of installs
             color='Type',             # Optional: color by Type
             points='all',             # Show all points as jitter
             log_y=True,               # Optional: log scale for better visualization
             title='Distribution of Installs: Free vs Paid Apps')

fig.update_layout(yaxis_title='Number of Installs (log scale)',
                  xaxis_title='App Type')

fig.show()

# Plotly Box Plots: Revenue by App Category


In [194]:

fig = px.box(df_apps_cleaned,
             x='Category',                 # App categories
             y='Revenue_Estimate',         # Revenue estimate
             color='Category',             # Optional: color by category
             points='all',                 # Show individual points (outliers)
             log_y=True,                   # Optional: log scale for better visualization
             title='Revenue Distribution by App Category')

fig.update_layout(xaxis_title='App Category',
                  yaxis_title='Estimated Revenue (log scale)')

fig.show()

# How Much Can You Charge? Examine Paid App Pricing Strategies by Category



In [195]:
paid_apps = df_apps_cleaned[df_apps_cleaned['Type'] == 'Paid']
median_price = paid_apps['Price'].median()
print("Median price of paid apps:", median_price)


Median price of paid apps: 2.99


# Conclusion and Findings

Based on our analysis of the Google Play Store app data, we've gained several key insights:

*   **Highly Competitive Categories:** Categories like 'FAMILY' and 'GAME' have the highest number of apps, indicating high competition.
*   **Most Popular Categories:** 'GAME', 'COMMUNICATION', and 'TOOLS' have the highest total downloads, suggesting strong market demand.
*   **Paid vs. Free Apps:** The box plot clearly shows that free apps have a significantly higher median number of installations compared to paid apps. This highlights the potential loss in downloads when opting for a paid model.
*   **Revenue Potential:** While the median price of paid apps is relatively low ($2.99), some apps in categories like 'FAMILY', 'LIFESTYLE', and 'FINANCE' have generated substantial estimated revenue. However, the distribution of revenue is highly skewed, with a few apps earning the majority.
*   **Highest Grossing Apps:** The analysis of estimated revenue revealed that popular games and some "I am rich" type apps (despite their high price and low installs) are among the highest earners. This suggests that both high volume (low price, high installs) and niche markets (high price, low installs) can be profitable.
*   **App Size:** The largest apps are around 100MBs, which might indicate a practical limit or user preference for app size.
*   **Highest Rated Apps:** While many apps have a perfect 5.0 rating, many of these have a very low number of reviews, suggesting that a high rating alone isn't a strong indicator of widespread quality or popularity.

Overall, the analysis provides a data-driven perspective on the dynamics of the Google Play Store, offering valuable insights for developers considering building and marketing their own apps.